tests/functional/test_cache.py

```diff
@@ -0,0 +1,101 @@
+import os
+import shutil
+
+
+def _cache_dir(script):
+    result = script.run(
+        'python', '-c',
+        'from pip._internal.locations import USER_CACHE_DIR;'
+        'print(USER_CACHE_DIR)'
+    )
+    return result.stdout.strip()
+
+
+def test_cache_info(script, monkeypatch):
+    result = script.pip('cache', 'info')
+    cache_dir = _cache_dir(script)
+
+    assert 'Location: {}'.format(os.path.normcase(cache_dir)) in result.stdout
+    # TODO(@duckinator): This should probably test that the number of
+    #   packages is actually correct, but I'm not sure how to do that
+    #   without pretty much re-implementing the entire cache info command.
+    assert 'Packages: ' in result.stdout
+
+
+def test_cache_list(script, monkeypatch):
+    cache_dir = _cache_dir(script)
+    wheel_cache_dir = os.path.join(cache_dir, 'wheels')
+    destination = os.path.join(wheel_cache_dir, 'arbitrary', 'pathname')
+    os.makedirs(destination)
+    with open(os.path.join(destination, 'yyy-1.2.3.whl'), 'w'):
+        pass
+    with open(os.path.join(destination, 'zzz-4.5.6.whl'), 'w'):
+        pass
+    result = script.pip('cache', 'list')
+    assert 'yyy-1.2.3' in result.stdout
+    assert 'zzz-4.5.6' in result.stdout
+    shutil.rmtree(os.path.join(wheel_cache_dir, 'arbitrary'))
+
+
+def test_cache_list_too_many_args(script, monkeypatch):
+    script.pip('cache', 'list', 'aaa', 'bbb',
+               expect_error=True)
+
+
+def test_cache_list_with_pattern(script, monkeypatch):
+    cache_dir = _cache_dir(script)
+    wheel_cache_dir = os.path.join(cache_dir, 'wheels')
+    destination = os.path.join(wheel_cache_dir, 'arbitrary', 'pathname')
+    os.makedirs(destination)
+    with open(os.path.join(destination, 'yyy-1.2.3.whl'), 'w'):
+        pass
+    with open(os.path.join(destination, 'zzz-4.5.6.whl'), 'w'):
+        pass
+    result = script.pip('cache', 'list', 'zzz')
+    assert 'yyy-1.2.3' not in result.stdout
+    assert 'zzz-4.5.6' in result.stdout
+    shutil.rmtree(os.path.join(wheel_cache_dir, 'arbitrary'))
+
+
+def test_cache_remove(script, monkeypatch):
+    cache_dir = _cache_dir(script)
+    wheel_cache_dir = os.path.join(cache_dir, 'wheels')
+    destination = os.path.join(wheel_cache_dir, 'arbitrary', 'pathname')
+    os.makedirs(destination)
+    with open(os.path.join(wheel_cache_dir, 'yyy-1.2.3.whl'), 'w'):
+        pass
+    with open(os.path.join(wheel_cache_dir, 'zzz-4.5.6.whl'), 'w'):
+        pass
```

Comment:
Make this a fixture?

Reply:
Agreed, but I'm not sure how fixtures work in the test setup pip uses, so focusing on the things I can get done quickly first. :+1:

Reply:
+1 on not worrying about this now.

Reply:
I've not turned it into a fixture yet, but I have extracted it out into a function. So it's at least a lot less repetitive!

Reply:
Converted it into a fixture. It may need more work, but it's definitely improved things significantly. :+1: